infrastructure/src/main/java/org/corfudb/infrastructure/BatchProcessor.java

```diff
@@ -114,127 +102,152 @@ private void recordRunnable(Runnable fsyncRunnable, Optional<Timer> fsyncTimer)
                                         tails = new TailsResponse(streamLog.getLogTail());
                                         break;
-
-                                    case TailsRequest.STREAMS_TAILS:
-                                        tails = streamLog.getTails(tailsRequest.getStreams());
+                                    case STREAMS_TAILS:
+                                        tails = streamLog.getTails(currentOp.getRequest()
+                                                .getPayload()
+                                                .getTailRequest()
+                                                .getStreamList()
+                                                .stream()
+                                                .map(CorfuProtocolCommon::getUUID)
+                                                .collect(Collectors.toList()));
                                         break;
-
                                     default:
                                         tails = streamLog.getAllTails();
                                         break;
                                 }
 
                                 tails.setEpoch(sealEpoch);
-                                currOp.setResultValue(tails);
+                                currentOp.setResultValue(tails);
                                 break;
                             case LOG_ADDRESS_SPACE_QUERY:
                                 // Retrieve the address space for every stream in the log.
                                 StreamsAddressResponse resp = streamLog.getStreamsAddressSpace();
                                 resp.setEpoch(sealEpoch);
-                                currOp.setResultValue(resp);
+                                currentOp.setResultValue(resp);
                                 break;
                             default:
-                                log.warn("Unknown BatchWriterOperation {}", currOp);
+                                log.warn("batchWriteProcessor: unknown operation {}", currentOp);
                         }
                     } catch (Exception e) {
-                        log.error("Stream log error. Batch [queue size={}]. StreamLog: [trim mark: {}].",
-                                operationsQueue.size(), streamLog.getTrimMark(), e);
-                        currOp.getFutureResult().completeExceptionally(e);
+                        log.error("batchWriteProcessor: stream log error. Batch: [queue size={}]. " +
+                                "StreamLog: [trim mark={}].", operationsQueue.size(), streamLog.getTrimMark(), e);
+
+                        currentOp.getFutureResult().completeExceptionally(e);
                     }
-                    res.add(currOp);
 
-                    processed++;
-                    lastOp = currOp;
+                    res.add(currentOp);
+                    numProcessed++;
+                    lastOp = currentOp;
```

Comment:
Error handling logic here is not correct. If an exception happens within a batch then all requests should be errored out because the write/fsync won't guarantee any boundaries.

Reply:
Actually, IO errors are fatal in this cause and the whole LogUnit should be re-initialized.